### PR TITLE
[Windows] Add initial implementation for TaskQueue

### PR DESCRIPTION
**Pull request by alexander-shaposhnikov**
Add initial implementation for TaskQueue.

### CODE REVIEW

**Comment by compnerd** [comment]
CC: @jrose-apple

**Comment by compnerd** [comment]
@swift-ci please test

**Review by jrose-apple — COMMENTED** [review]
A lot of this looks shared with the Unix implementation. Can that be factored out at all? Even if it means including it into the dummy implementation and letting it go unused.

**Comment by compnerd** [comment]
Please run the test suite - you need to update the test suite for this change - Driver.batch_mode_aux_file_order and Driver.batch_mode need to be updated to reflect the fact that they are now supported on Windows as well.

File: lib/Basic/Windows/TaskQueue.inc [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+    if (!RC) {
+      SetLastError(Err);
+      // -2 indicates a crash or timeout as opposed to failure to execute.
+      // See the comments on llvm::sys::Wait for more details.
+      T->PI.ReturnCode = -2;
+    }
+    // Convert the return code in the same way as llvm::sys::Wait does.
+    if (Status != 0) {
+      // Pass 10(Warning) and 11(Error) to the callee as negative value.
+      if ((Status & 0xBFFF0000U) == 0x80000000U)
```
[the remark this answers is not in the document]
Comment (alexshap) [reply lib/Basic/Windows/TaskQueue.inc]: i took a look at where this was introduced in the first place - https://github.com/llvm-mirror/llvm/commit/5d2f8c31556b6a5ce29d2cad314bc48af879c5b3 
one can also have a look at https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-erref/596a1078-e883-4972-9bbc-49e60bebca55  (to see what kind of status values belong to these ranges (0x8000XXXX and 0xC000XXXX)), but I'm not sure I can add much clarity here (tbh i kind of not sure about the exact reason why llvm::sys::wait does this conversion at the moment). I can add FIXME or TODO to indicate that this needs to be figured out.

**Comment by compnerd** [comment]
@swift-ci please test

File: lib/Basic/Windows/TaskQueue.inc (hunk 1 above)
[the remark this answers is not in the document]
Comment (jrose-apple) [reply lib/Basic/Windows/TaskQueue.inc]: Hmm. I was hoping there'd be a constant in a Windows header that defined at least some of these. If not, I guess referencing that MS-ERREF page might be the next best thing.

[the remark this answers is not in the document]
Comment (jrose-apple) [reply lib/Basic/Windows/TaskQueue.inc]: Ah, the top-level page is probably a little clearer if we have to resort to that: https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-erref/87fba13e-bf06-450e-83b1-9241dc81e781

[the remark this answers is not in the document]
Comment (alexshap) [reply lib/Basic/Windows/TaskQueue.inc]: done

**Comment by compnerd** [comment]
@swift-ci please test

File: lib/Basic/Windows/TaskQueue.inc [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+unsigned TaskQueue::getNumberOfParallelTasks() const {
+  return NumberOfParallelTasks;
+}
+
+void TaskQueue::addTask(const char *ExecPath, ArrayRef<const char *> Args,
+                        ArrayRef<const char *> Env, void *Context,
+                        bool SeparateErrors) {
+  std::unique_ptr<Task> T(
+      new Task(ExecPath, Args, Env, Context, SeparateErrors));
+  QueuedTasks.push(std::move(T));
```
Comment (compnerd) [inline lib/Basic/Windows/TaskQueue.inc]: This could probably be written using `emplace_back` and `make_unique`.

**Review by compnerd — COMMENTED** [review]
@jrose-apple - any other thoughts?  This makes a significant impact on the Windows build, so would be nice to get this merged.

**Review by jrose-apple — COMMENTED** [review]
Okay, looks good to me.

File: lib/Basic/Default/Task.inc [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+  bool execute();
+};
+
+bool Task::execute() {
+  llvm::SmallVector<const char *, 128> Argv;
+  Argv.push_back(ExecPath);
+  Argv.append(Args.begin(), Args.end());
+  Argv.push_back(nullptr);
+
+  llvm::Optional<llvm::ArrayRef<llvm::StringRef>> Envp =
```
[the remark this answers is not in the document]
Comment (alexshap) [reply lib/Basic/Default/Task.inc]: will do)

File: lib/Basic/Default/TaskQueue.inc [hunk 4]
```diff
@@ -84,7 +55,8 @@ unsigned TaskQueue::getNumberOfParallelTasks() const {
 void TaskQueue::addTask(const char *ExecPath, ArrayRef<const char *> Args,
                         ArrayRef<const char *> Env, void *Context,
                         bool SeparateErrors) {
-  std::unique_ptr<Task> T(new Task(ExecPath, Args, Env, Context, SeparateErrors));
+  std::unique_ptr<Task> T(
+      new Task(ExecPath, Args, Env, Context, SeparateErrors));
   QueuedTasks.push(std::move(T));
```
[the remark this answers is not in the document]
Comment (jrose-apple) [reply lib/Basic/Default/TaskQueue.inc]: I'm not sure that's easier to read.

**Comment by alexshap** [comment]
@swift-ci please test

**Comment by gmittert** [comment]
@swift-ci please test

**Comment by alexshap** [comment]
@swift-ci please test

**Pull request merged**